cpp/src/hash/concurrent_unordered_map.cuh

```diff
@@ -311,6 +311,46 @@ class concurrent_unordered_map {
     }
   }
 
+  /**---------------------------------------------------------------------------*
+   * @brief Searches the map for the specified key using different hash and equality functions.
+   *
+   * This find function enables creation of the map and querying of
+   * the map to use different functions.  This is useful in situations
+   * where we use indirect access to get to elements being compared,
+   * and want to have different tables used during query.
+   *
+   * @note `find` is not threadsafe with `insert`. I.e., it is not safe to
+   *do concurrent `insert` and `find` operations.
+   *
+   * @param k The key to search for
+   * @param f_hash hashing function for this find
+   * @param f_equal equality function for this find
+   * @return An iterator to the key if it exists, else map.end()
+   *---------------------------------------------------------------------------**/
+  __device__ const_iterator find(key_type const& k, hasher f_hash,
```

Comment:
I really want to eliminate these functions where you specify the hash and equality as function parameters. See https://github.com/rapidsai/cudf/issues/3518

They should be specified at the object level of `concurrent_unordered_map`. I'm very skeptical about the need to have different notions of hashing and equality. It was my mistake in the `multimap` class to have `insert/find` functions that look like this.

Reply:
I guess my use case was to insert using indirection from one table and find (probe) using indirection from another table.  In this case the hashing function needs to be able to compute the hash using rows from a different table and the find needs to be able to compare entries from the new table to entries in the old table.

The draft join kernels (and I assume the legacy ones are similar) work around this by creating a probe kernel that essentially implements its own find function outside of the multimap (totally dependent on an intimate understanding of the implementation of multimap).  Of course, the regular join kernels need to enumerate all matches (and they use the multimap where you can have multiple matches to your key), so the code needs to be more sophisticated anyway.

This was an attempt to create a direct mechanism in the concurrent_unordered_map to find exactly what you want.  This find exactly computes the single entry that you want and returns either an iterator point to it or this->end().  It greatly simplifies the use of the concurrent_unordered_map for the use case of insert things indirectly from table 1, find things indirectly from table 2.

The same technique could be applied to the concurrent_unordered_multimap by creating a more sophisticated iterator where ++ automatically skipped over rows that don't match.

I could move this logic into the create function/constructor, and identify functions for insertion and functions for probing (which would default to the same).  That would perhaps be a little cleaner.

Alternatively, I can eliminate this and just write a more sophisticated kernel function to find matches.  It just felt like this was a common use case and could be addressed directly.

Reply:
OK, just pushed an update where I specified these at the object level.  Existing create method returns an object that behaves exactly as today.  Added a new create method that allows you to specify a different hash and equality operator for use in the find call.  These leaves us with a cleaner find implementation.